ESQL: Improve grammar to allow identifiers with .

costin opened the pull request:
Extend the unquoted identifier to contain . not just numbers. Without it
 the lexer picks the characters as decimal literal which leads to errors

Fix #100312

costin committed:
ESQL: Improve grammar to allow identifiers with . and numbers

Extend the unquoted identifier to contain . not just numbers. Without it
 the lexer picks the characters as decimal literal which leads to errors

Fix #100312

elasticsearchmachine commented:
Hi @costin, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

costin committed:
Update docs/changelog/100740.yaml

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

costin commented:
The issue with unquoted field names that contain `.` is we are already using that for qualifying names:
`a.b` means field b inside a.  Similar to the JSON object notation `a.b.c`.
Under this criterion `a.1m.4321` gets parsed as field `1m` under `a` however we do not allow unquoted field names to start with digits because of decimal notation, e.g. `1e9` is a field name or a number?
The improvement in this PR adds `.` to the unquoted identifiers so that `a.1m` becomes just one field name - there are no qualifiers.
However the test now fails for names such as `a.@b` (e.g `source.@timestamp`) because `@` appears inside the field name...

To make matters worse we have different lexer rules between different commands hence why the same unquoted identifier works in one but fails in the other - in once is just a string, in the other it's an expression.
Before going down this path, we need to figure out the handling of qualified names.

astefan commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -127,12 +131,15 @@ PERCENT : '%';
 
 UNQUOTED_IDENTIFIER
-    : LETTER (LETTER | DIGIT | '_')*
+    : LETTER UNQUOTED_ID_BODY*
```
This will allow digits inside "subfields", but not on the root field. For example `123elasticsearch.node.stats.os.cpu.load_avg.1m`

alex-spies replied:
Do we want to allow unquoted identifiers that start with digits, though? If so, we'd at least have to disallow identifiers being all digits, no?

bpintea replied:
> we'd at least have to disallow identifiers being all digits, no?

ES accepts both fields `"123"` and `"123.456"`.

costin replied:
Which we accept as long as the fields are quoted "123" - same with field above "a.1m.4321" is accepted; the problem is handling fields which are NOT quoted.

alex-spies commented on docs/changelog/100740.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 100740
+summary: "ESQL: Improve grammar to allow identifiers with"
```
Changelog didn't like the period.

alex-spies reviewed (approved):
LGTM

alex-spies commented:
I wonder if this change affects lexing/parsing of multi-segment identifiers that contain quoted segments, e.g.:
```
asdf.`1234`.fdsa
```
Currently this should be allowed and fine, but this PR could break cases like this, so we should double check if this is covered by tests. (Not familiar enough with our parsing/lexing to say.)

An alternative would be to use a (negative) lookahead in the lexer via a semantic predicate (like in this [SO post](https://stackoverflow.com/questions/24194110/antlr4-negative-lookahead-in-lexer)), to allow an identifier to start with a digit only if preceded by a `DOT` fragment. However, that might be problematic given that we generate both Java and JavaScript code from the same antlr4 definition :/

costin committed:
Update ANTLR files

costin commented:
I've spent some time on the issue and we essentially have to pick whether we align all commands to have the same identifier definition (and restrictions) or keep things a bit loose which results in problems like in the issue above.
While I like the flexibility and I see that we took advantage of it in our tests, I'm opting for the latter since it less confusing.

To restate the problem, right now from, drop, keep, rename and enrich commands due to their simpler definition, accept a much larger set of identifiers to be declared _without_ quotes.
To restate, the issue we're debating here is dealing with unquoted identifiers which, I would argue, should be discouraged form.

1. fields with docs
```
from index
| where `duration.1m` > 1    // name needs quoting because of .
| keep duration.1m              // name can be unquoted 
```

2. field with parenthesis
```
from index
| stats count(*), max(field)   
| where `count(*)` > 1          // aliased function name need to be quoted
| keep count(*)                    // same alias can be used unquoted
```

3. fields with special chars such as @ or +
```
from index
| where `@domain+1` > 1   // needs quoting
| keep @domain+1            // used unquoted
```

I find case 1 the most concerning since `.` is used for qualified names `a. b`. We currently don't have plugged in but we will especially when dealing with different sources. This means supporting qualifiers both in unquoted and quoted fields (and having to deal with escaping among other things).

2. is inconvenient when trying to work with things right after a stats however considering there's an alias version I don't think that's much of a problem.

3.  to me, this is an example why we'd want the identifiers to be the same everywhere else. The expression `@domain+1` has different meaning depending on the command - in where it's an expression while in keep is a field as whole.

As a user, I find 3 quite trappy for no real advantage.

elasticsearchmachine commented:
Hi @costin, I've updated the changelog YAML for you.

alex-spies commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -64,10 +68,27 @@ fragment EXPONENT
     ;
 
+fragment UNQUOTED_ID_BODY
+    : (LETTER | DIGIT | UNDERSCORE)
```
++ to removing the dot again from the unquoted identifier. I agree that dots will likely need to retain special meaning to navigate nested objects. If I understand correctly, this should also bring back the ability to quote just part of an identifier, e.g. 
```
foo.bar.`1234asdf`
```

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser.java:
Removal of the generated files looks accidental; does not currently compile without manually building the parser first.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -733,7 +733,7 @@ public void testMissingInputParams() {
 
     public void testFieldContainingDotsAndNumbers() {
-        LogicalPlan where = processingCommand("where a.b.1m.4321 ");
+        LogicalPlan where = processingCommand("where `a.b.1m.4321`");
```
suggestion: IMHO it'd be good to also add a test case for
```
where a.b.`1m`.`4321`
```
or similar.

alex-spies reviewed (approved):
This looks like we still need to make the CI happy (resp. make it compile), but the proposed refactoring LGTM.

alex-spies commented:
> I've spent some time on the issue and we essentially have to pick whether we align all commands to have the same identifier definition (and restrictions) or keep things a bit loose which results in problems like in the issue above. While I like the flexibility and I see that we took advantage of it in our tests, I'm opting for the latter since it less confusing.
> 
> To restate the problem, right now from, drop, keep, rename and enrich commands due to their simpler definition, accept a much larger set of identifiers to be declared _without_ quotes. To restate, the issue we're debating here is dealing with unquoted identifiers which, I would argue, should be discouraged form.
> 
> 1. fields with docs
> 
> ```
> from index
> | where `duration.1m` > 1    // name needs quoting because of .
> | keep duration.1m              // name can be unquoted 
> ```
> 
> 2. field with parenthesis
> 
> ```
> from index
> | stats count(*), max(field)   
> | where `count(*)` > 1          // aliased function name need to be quoted
> | keep count(*)                    // same alias can be used unquoted
> ```
> 
> 3. fields with special chars such as @ or +
> 
> ```
> from index
> | where `@domain+1` > 1   // needs quoting
> | keep @domain+1            // used unquoted
> ```
> 
> I find case 1 the most concerning since `.` is used for qualified names `a. b`. We currently don't have plugged in but we will especially when dealing with different sources. This means supporting qualifiers both in unquoted and quoted fields (and having to deal with escaping among other things).
> 
> 2. is inconvenient when trying to work with things right after a stats however considering there's an alias version I don't think that's much of a problem.
> 3. to me, this is an example why we'd want the identifiers to be the same everywhere else. The expression `@domain+1` has different meaning depending on the command - in where it's an expression while in keep is a field as whole.
> 
> As a user, I find 3 quite trappy for no real advantage.

I think identifiers should be treated consistently across all clauses; I think for users it will be confusing that they need to do less quoting in `KEEP` than in `WHERE`. If I understand correctly, this could also simplify our lexing, considering that it requires lots of different modes at the moment.

dej611 commented:
In Kibana we're about to merge some logic that heavily rely on the grammar definitions to run validation/autocomplete/etc... which syncs the Kibana and ES grammar into a single version (Kibana used a fork of it): https://github.com/elastic/kibana/pull/170071

In case changes like these will be merged into ES it would be helpful to provide enough time to sync with these changes as well and not break everything on the Kibana side.

costin committed:
Merge branch 'main' into fix/100312

costin committed:
Update grammar and check CSV tests

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -139,9 +139,9 @@ emp_no:integer | birth_date:date               | x:date
 evalDateTruncGrouping
-from employees | eval y = date_trunc(1 year, hire_date) | stats count(emp_no) by y | sort y | keep y, count(emp_no) | limit 5;
+from employees | eval y = date_trunc(1 year, hire_date) | stats c = count(emp_no) by y | sort y | keep y, c | limit 5;
 
-y:date                        | count(emp_no):long
+y:date                        | c:long
```
A side effect of this change - the expression has to be quoted or an alias used.

luigidellaquila replied:
I think this is a positive side effect; being able to use an unquoted function-like alias seems inconsistent

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -1,24 +1,24 @@
-STATS : 'stats' -> pushMode(EXPRESSION);
-WHERE : 'where' -> pushMode(EXPRESSION);
-UNKNOWN_CMD : ~[ \r\n\t[\]/]+ -> pushMode(EXPRESSION);
+DISSECT : 'dissect'           -> pushMode(EXPRESSION_MODE);
```
The main change - commands are split to have their own mode; due to their particularities there are about 7 modes...

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -31,16 +31,20 @@ MULTILINE_COMMENT
+//
 mode EXPLAIN_MODE;
-EXPLAIN_OPENING_BRACKET : '[' -> type(OPENING_BRACKET), pushMode(DEFAULT_MODE);
-EXPLAIN_PIPE : '|' -> type(PIPE), popMode;
+EXPLAIN_OPENING_BRACKET : OPENING_BRACKET -> type(OPENING_BRACKET), pushMode(DEFAULT_MODE);
+EXPLAIN_PIPE : PIPE -> type(PIPE), popMode;
 EXPLAIN_WS : WS -> channel(HIDDEN);
 EXPLAIN_LINE_COMMENT : LINE_COMMENT -> channel(HIDDEN);
 EXPLAIN_MULTILINE_COMMENT : MULTILINE_COMMENT -> channel(HIDDEN);
```
Used this approach elsewhere in trying to reuse the existing token declarations.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -150,31 +172,28 @@ EXPR_MULTILINE_COMMENT
+//
+// FROM command
+//
+mode FROM_MODE;
+FROM_PIPE : PIPE -> type(PIPE), popMode;
+FROM_OPENING_BRACKET : OPENING_BRACKET -> type(OPENING_BRACKET), pushMode(FROM_MODE), pushMode(FROM_MODE);
+FROM_CLOSING_BRACKET : CLOSING_BRACKET -> type(CLOSING_BRACKET), popMode, popMode;
+FROM_COMMA : COMMA -> type(COMMA);
+FROM_ASSIGN : ASSIGN -> type(ASSIGN);
```
Type reuse and consistency.

alex-spies replied:
Why do we need to push/pop `FROM_MODE` twice?

astefan replied:
I think it's because of the double push the opening_bracket is doing: https://github.com/elastic/elasticsearch/pull/100740/files#diff-e78771d02134981fe1563249a0fbbdabf95b40299b2086dd19fccef6e87f39c4R150
Same for closing_bracket.

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.g4:
```diff
@@ -189,3 +208,143 @@ SRC_MULTILINE_COMMENT
+    : (LETTER | DIGIT | UNDERSCORE | ASTERISK)
+    ;
+
+PROJECT_UNQUOTED_IDENTIFIER
+    : (LETTER | ASTERISK) UNQUOTED_ID_BODY_WITH_PATTERN*
+    | (UNDERSCORE | ASPERAND) UNQUOTED_ID_BODY_WITH_PATTERN+
+    ;
```
Allow * in unused field declaration in keep, drop and project...

costin commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -163,20 +171,20 @@ orderExpression
     ;
 
 keepCommand
-    :  KEEP sourceIdentifier (COMMA sourceIdentifier)*
-    |  PROJECT sourceIdentifier (COMMA sourceIdentifier)*
+    :  KEEP qualifiedNamePattern (COMMA qualifiedNamePattern)*
+    |  PROJECT qualifiedNamePattern (COMMA qualifiedNamePattern)*
```
Support nested field without quoting by using qualified names.
That is treating height.float as height - one unquoted identifier and float another unquoted identifier.

costin committed:
Add missing classes

costin committed:
Spotless...

costin committed:
Fix yaml test

dej611 commented:
I've migrated our grammar to this new one and fixed most of the token renames: https://github.com/elastic/kibana/pull/172148
The test suite found out a regression with the new grammar, in particular related to the `-` within identifiers:
```esql
from a | enrich my-policy
```

That used to be valid before, while now it throws the following syntax errors:
```
SyntaxError: token recognition error at: '-'
SyntaxError: extraneous input 'pol*' expecting <EOF>
```

Other than that, in general syntax errors are way more noisy:
* `show info something` passed from a single syntax error (`SyntaxError: extraneous input 'something' expecting <EOF>`) to 9 now (one for each char)
 * `from index | keep 4.5` passed from 2 to 4 syntax errors
 * `from a | rename fn() as a` passed from 0 to 2 syntax errors
 * etc... (can find more in the linked PR)

alex-spies commented:
> I've migrated our grammar to this new one and fixed most of the token renames: [elastic/kibana#172148](https://github.com/elastic/kibana/pull/172148) The test suite found out a regression with the new grammar, in particular related to the `-` within identifiers:
> 
> ```
> from a | enrich my-policy
> ```

Just checked and with this PR's code, this regression is in fact inconsistent with `from`; the following works:

```
from my-idx | keep whatever
```

There's another "inconsistency" with how field names are handled:
```
from my-idx | eval x = my-field
```
This parses as `my SUB field`; users need backticks around `my-field` to make this work.
But field names are not index/policy names, so maybe this inconsistency is not so bad.

astefan reviewed (approved):
LGTM
Documentation will need an update, as well. CC @bpintea @abdonpijpelink

luigidellaquila reviewed (approved):
LGTM

costin commented:
> I've migrated our grammar to this new one and fixed most of the token renames: [elastic/kibana#172148](https://github.com/elastic/kibana/pull/172148)
> The test suite found out a regression with the new grammar, in particular related to the `-` within identifiers:

> Just checked and with this PR's code, this regression is in fact inconsistent with from; the following works:

I think for enrich policy name, we can improve the grammar to have the same lexing as in `from` - that is allow - and other special characters in its name so there's no need for quoting.

> Other than that, in general syntax errors are way more noisy:
>
> show info something passed from a single syntax error (SyntaxError: extraneous input 'something' expecting <EOF>) to 9 now (one for each char)
> from index | keep 4.5 passed from 2 to 4 syntax errors
> from a | rename fn() as a passed from 0 to 2 syntax errors
> etc... (can find more in the linked PR)

That's something that we could look into improving however it's a fragile approach since the error messages depend on the grammar which generates different ANTLR internal representation.  Which is triggered by a change like the above.

`rename fn()` doesn't work anymore since fn() needs to be a field identifier meaning `()` cannot be used unquoted. The fix is to add backticks: ``rename `fn()` ``

astefan commented:
Leaving this one here, since I don't see a solution to it:`` row `my-field`=123 | stats count(`my-field`) | eval x = `count(`my-field`)` ``

costin committed:
Update parsing of enrich policy name

costin commented:
Updated the grammar to allow richer policies than identifiers - due to lexing that ended up more complicated than expected having to add a submode for enrich since otherwise the lexing between from identifiers and field identifiers tripped each other.

> Leaving this one here, since I don't see a solution to it:row `my-field`=123 | stats count(`my-field`) | eval x = `count(`my-field`)`

added a test for it: ``stats count(`my-field`) | keep `count(my-field)` ``

elasticsearchmachine commented:
Hi @costin, I've updated the changelog YAML for you.

costin committed:
Update docs/changelog/100740.yaml

costin committed:
Merge branch 'main' into fix/100312

costin committed:
Address feedback on field name quoting

costin commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/StatementParserTests.java:
```diff
@@ -799,6 +800,29 @@ public void testMissingInputParams() {
+
+    public void testQuotedName() {
+        // row `my-field`=123 | stats count(`my-field`) | eval x = `count(`my-field`)`
+        LogicalPlan plan = processingCommand("stats count(`my-field`) |  keep `count(my-field)`");
```
Good catch - thanks for this.
This was incorrect  - the correct syntax should be 
```
keep `count(``my-field``)`  ; the backquotes are escaped
```
since we want to differentiate between
```
stats count(a - b)  ; which we want to support at some point
keep `count(a-b)`  

stats count(`a-b`)   ; this is not an expression but a field name hence the ` need to be kept around meaning
| keep `count(``a``-``b``)`

; this means any backquote from the user is kept as is since we're using the query verbatim

stats count(`a`-`b`)         ; need to preserve the quotes
| keep `count(``a``-``b``)` ; 
```

I've added a CSV test and caught a bug in our identifier handling.
I've tried to play with removing redundant quotes however as a user I found the behavior surprising.

astefan replied:
Thanks @costin. This really LGTM.

costin committed:
Skip test due to breaking change

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:
```diff
@@ -783,7 +783,7 @@ FROM sample_data
-fieldEscaping
+fieldEscaping#[skip:-8.12.99, reason:Fixed bug in 8.13 of removing the leading/trailing backquotes of an identifier]
 FROM sample_data
 | stats count(`event_duration`) |  keep `count(``event_duration``)`
```
@abdonpijpelink This PR introduces a (subtle) breaking change due to a previous bug:
- the leading/trailing backquotes from an identifier are removed which means that if folks used them (rare case) they would have to update their query.
```
stats count(`event_duration`) produces the alias count(``event_duration`) - same text
```
Since 8.13 (this PR) keep needs to use the same rules for quoting as the rest of commands meaning that field names with special characters need to be quoted and the backquote itself be escaped through repetition.
So 
```
count(`event_duration`)  becomes `count(``event_duration``)`
```
However that is not the case in 8.12 - the grammar rules are slightly different so the text is interpreted verbatim meaning the escaping of backticks is taken as is hence why the field is not found:
```
 Unknown column [count(``event_duration``)], did you mean [count(`event_duration`)]?
```

We need to put this information somewhere (not sure where) and in a smaller format.

abdonpijpelink commented:
I propose we change the changelog yaml `summary` on this PR into something like: 

```
Referencing expressions that contain backticks requires <<esql-identifiers,escaping those backticks>>.
```

And in parallel, we update the [Identifiers section on the Syntax page](https://www.elastic.co/guide/en/elasticsearch/reference/master/esql-syntax.html#esql-identifiers) for 8.13 to include the new behavior.

costin committed:
Update summary
